### ISSUE TITLE
Use threads to visit directories and parse files in parallel in SourceRoot

### ISSUE THREAD

**Issue by jimshowalter** [javaparser/javaparser#1046]
We're getting:

[info] Parsed 26810 files
[info] ELAPSED: 70 (383 files/second)

We wrote a regex-based tool that isn't as powerful as a Java compiler front-end, but still does a lot of work, and with that we're getting:

26707 all Java files
32581 total files
[info] ELAPSED: 19 (1714 files/second)

It's likely that much of the difference in speed is due to the 1000 threads and 500 open files allowed in the faster one.

**Comment by ftomassetti** [comment]
Well, a regex-based tool does probably a fraction of the work of JavaParser so I would say it is a pretty good comparison for JavaParser

**Comment by matozoid** [comment]
Good suggestion though - this would be a nice one for a user contribution since the issue is constrained to one file.

**Comment by jimshowalter** [comment]
I hope to have a user contribution for you soon.

Regarding "a pretty good comparison for JavaParser", don't sell yourselves short. The parser is fine. But you're running it synchronously. On an 8-way box, expect to see a corresponding 8x improvement in speed.

Our Jenkins server runs on a 48-core box.

Threads are good.

**Comment by jimshowalter** [comment]
It's ironic that you labelled this easy, because I just blew six hours on it. Everything I tried just made it slower. It threaded, and it wrote out the number of parses done in background vs. synchronous (it throttled), but the thread performance was abysmal compared to no threading. I used the same pattern we've used numerous times in other tools at my company.

**Comment by matozoid** [comment]
:-D yes, it's one of the worst labels we have. It should be "a good first contribution" or so. We had more easy tasks explode like this

**Comment by ryan-beckett** [comment]
@jimshowalter: What do you, "synchronously"? Single-threaded?

__Edit:__ Nevermind, I understand.

**Comment by jimshowalter** [comment]
The idea is to visit directories on separate threads and parse files on separate threads, up to a tunable limit. I wrote a tool that does that for a different application. The trick is to avoid needing to synchronize, and the way to do that is to make each visited directory a separate object that contains a set of files parse in that directory, and a set of subdirectories. At the end, you visit the tree and collect all parse files. This took our parsing from minutes to seconds.

Unfortunately, when I tried applying the same approach using javaparser, it made things worse. It's possible javaparser consumes so much memory that trying to background it is doomed.

**Comment by matozoid** [comment]
That reminded me to build this thing: https://github.com/javaparser/javaparser/pull/1069 :-)

**Comment by ryan-beckett** [comment]
Or another idea is to leave the visit/parse synchronous, but break up each task into worker threads.

> It's possible javaparser consumes so much memory that trying to background it is doomed.

Did you profile the memory?

**Comment by jimshowalter** [comment]
I looked at memory consumption and it was around 7 GB, but there was 14 available. It was acting more like CPU-bound.

**Comment by ryan-beckett** [comment]
> I looked at memory consumption and it was around 7 GB, but there was 14 available. It was acting more like CPU-bound.

@jimshowalter: Wouldn't this be an IO-bound issue? does this somehow correspond with your server hardware configuration 

For parsing ~12k files, I was able to go from ~356 secs to ~136 secs, with a negligible increase in memory, by putting the call to **JavaParser.parse** in the callback of a multi-threaded recursive tree walk class. If you want the code, let me know and I'll post it.

I'll work on trying to make this happen in **SourceRoot.java**.

**Comment by ryan-beckett** [comment]
@jimshowalter: Mind you I'm running an Athlon II X3.

**Comment by ryan-beckett** [comment]
@jimshowalter: I see what you mean. Using SourceRoot to parse many files uses way more memory!

**Comment by ryan-beckett** [comment]
@jimshowalter , @matozoid: I'm going to pass on this issue in favor of users using [this code](https://gist.github.com/ryan-beckett/f298ab6fe84f3fb8025aa4cb28b8c793) to parallelize parsing. It's uses a fraction of the memory and finishes way faster than using SourceRoot. I tried parallelizing SourceRoot, but it made the memory consumption worse.

**Comment by jimshowalter** [comment]
That was fast! And "I tried parallelizing SourceRoot, but it made the memory consumption worse." sounds very familiar.

**Comment by ryan-beckett** [comment]
@jimshowalter: What was fast? Yeah, I don't know what's going on with the SourceRoot. The purely recursive nature of the implementation explains the memory consumption, but I don't know why parallelizing it didn't help. I don't know the codebase well enough to bother trying to figure out. Perhaps it's the caching or relativizing paths... IDK.

**Comment by jimshowalter** [comment]
You implementing a solution was fast. (And the solution is fast.)

**Comment by ryan-beckett** [comment]
Ah, yes, but if I can only integrate it. I'm sure users would be quite happy! I'll take another stab at it tomorrow.

**Comment by ryan-beckett** [comment]
@matozoid: I've developed a working multi-thread source root, but there's an issue. Since **JavaParser** isn't thread safe, I have to either (1) create a new parser for each file, or (2) use a pool. (2) adds more complexity and I'd rather avoid it for now. (1) doubles the memory for **SourceRoot.parse(String, JavaParser, Callback)** for 3k files, but isn't an issue since the memory usage is ~30MB. (1) adds a little more memory for **SourceRoot.tryToParse**, but the caching is still the bottleneck. 

One idea is to copy the internal instance for each new parser that's made, but I don't see an API method to do that. Is it as simple as grabbing the configuration (i.e. **new JavaParser(oldParser.getConfiguration())**)? That would add a little CPU overhead, but it would still probably make the processing faster overall. Even if I implemented a pool, I'd still have to be able to copy the parser.

Synchronizing access to a single parser makes the implementation even slower.

As of now, the implementation creates a new thread for each directory during it's tree walk. I'm assuming any projects using this library would only use **SourceRoot** for processing a large number of files, and if that's not the case, the thread overhead is probably negligible.

Overall, using (1), I'm seeing a consistent 50% decrease in total processing time for a large number of files, so I think this is a good feature to have.

**TLDR**: I need to know the best way to create an instance of **JavaParser** from another instance.

**Comment by ryan-beckett** [comment]
@matozoid: As I'm using different sets of test data, I'm seeing that sometimes multi-threading isn't producing that much of a speed-up, which makes me wonder if it's a better idea to leave the **SourceRoot** class as is, and add a **AsynchronousSourceRoot** class to be used experimentally.

**Comment by ryan-beckett** [comment]
@matozoid: Another idea is to add my implementation as a new class **AsynchronousSourceRoot** to be used experimentally, but I still need a way to copy the parser.

**Comment by matozoid** [comment]
@ryan-beckett I have been thinking of splitting `SourceRoot` in two, and it sounds like this is the right time. My idea for the split was to have one class that focuses on quickly traversing a tree, where your stuff would end up, and the other would be the one that is more focused on live code modification, using the cache and `save()` etc.

Let me play around with this for a bit tonight.

Instance copying: I think you want the user to supply the `JavaParserConfiguration` and `new` a JavaParser with that configuration every time, but I'll check tonight.

**Comment by ryan-beckett** [comment]
> My idea for the split was to have one class that focuses on quickly traversing a tree, where your stuff would end up, and the other would be the one that is more focused on live code modification, using the cache and save() etc.

Doesn't this potentially break client code? If it ain't broke, don't try to fix it... I'm just trying to address the problem that @jimshowalter brought up. You're good with the architectural stuff.

I was hoping to either (1) do a PR for **SourceRoot** with the my implementation, or (2) do a PR for **AsynchronousSourceRoot** to be used experimentally. I say (2) because I just want to ensure end-users don't see a massive memory increase or decreased performance due to threading when it isn't needed (i.e. small number of files). I really don't think that's going to happen, but I'm being cautious. 

> Instance copying: I think you want the user to supply the JavaParserConfiguration and new a JavaParser with that configuration every time, but I'll check tonight.

I'll add **JavaParser.getParserConfiguration** to the API, and I'll submit a PR. And if you think it's OK, lets go with it. Otherwise, let me know your thoughts.

**Comment by ryan-beckett** [comment]
@matozoid: We can close this now.

**Comment by matozoid** [comment]
Yeah, I'll do the management stuff :-)
